### PR TITLE
Crawl Pipeline Fix

### PR DESCRIPTION
**Pull request by lokitoth**
Fix the Crawl pipeline's EntityLinking task, which breaks due to Cognitive Services discontinuing the individual endpoint for entity linking, merging it into Text Analytics.

### CODE REVIEW

**Comment by JohnLangford** [comment]
Great :-)

File: client/ClientDecisionService/ClientDecisionService.csproj [hunk 1]
```diff
@@ -133,20 +133,20 @@
     <Reference Include="Microsoft.CSharp" />
     <Reference Include="System.Data" />
     <Reference Include="System.Xml" />
-    <Reference Include="VowpalWabbit, Version=8.4.0.3, Culture=neutral, PublicKeyToken=a76afd1645210483, processorArchitecture=AMD64">
-      <HintPath>..\..\packages\VowpalWabbit.8.4.0.3\lib\net45\VowpalWabbit.dll</HintPath>
+    <Reference Include="VowpalWabbit, Version=8.6.0.1, Culture=neutral, PublicKeyToken=a76afd1645210483, processorArchitecture=AMD64">
```
Comment (rajan-chari) [inline client/ClientDecisionService/ClientDecisionService.csproj]: Nuget.org shows 8.4.0.3 [https://www.nuget.org/packages/VowpalWabbit/](https://www.nuget.org/packages/VowpalWabbit/)

File: client/ClientDecisionService/packages.config [hunk 2]
```diff
@@ -18,7 +18,8 @@
   <package id="Rx-Main" version="2.2.5" targetFramework="net45" />
   <package id="Rx-PlatformServices" version="2.2.5" targetFramework="net45" />
   <package id="System.Spatial" version="5.8.1" targetFramework="net45" />
-  <package id="VowpalWabbit" version="8.4.0.3" targetFramework="net452" />
-  <package id="VowpalWabbit.JSON" version="8.4.0.3" targetFramework="net452" />
   <package id="WindowsAzure.Storage" version="7.2.1" targetFramework="net45" />
+
+  <package id="VowpalWabbit" version="8.6.0.1" targetFramework="net452" />
```
Comment (rajan-chari) [inline client/ClientDecisionService/packages.config]: >8.6.0.1 [](start = 38, length = 7)

Nuget.org shows 8.4.0.3

File: Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs [hunk 3]
```diff
@@ -39,27 +39,35 @@ public static Task<HttpResponseMessage> Run(HttpRequestMessage req, TraceWriter
                     if (textBuilder.Length == 0)
                         return null;
 
-                    return Services.Limit(textBuilder.ToString(), 10240);
+                    string text = Services.Limit(textBuilder.ToString(), 10240);
```
Comment (rajan-chari) [inline Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs]: >10240 [](start = 73, length = 5)

Why 10k here but 5k in CreateRequestFromText()?

Comment (rajan-chari) [inline Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs]: >10240 [](start = 73, length = 5)

Does it make sense to introduce some constants here with additional documentation?  A link to spec perhaps?

File: client/ClientDecisionService/packages.config (hunk 2 above)
Comment (lokitoth) [reply client/ClientDecisionService/packages.config]: Ah, yeah, I should not include this until we can ship a new package publicly.

File: Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs (hunk 3 above)
Comment (lokitoth) [reply Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs]: Per Markus, this is due to ANSI vs UTF-16 character encoding (10k vs 5k per 10k byte buffer). Will add comment here, to make it clearer.

Comment (lokitoth) [reply Crawl/CognitiveServiceEntityLinking/CognitiveServiceEntityLinking.cs]: Yes.

**Comment by jackgerrits** [comment]
Did you mean to commit nuget.exe?

File: Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs [hunk 4]
```diff
@@ -24,6 +24,26 @@ static CognitiveServiceTextAnalytics()
             cogService = new CognitiveService("CogTextAnalytics", queryParams: "/sentiment");
         }
 
+        internal static TextAnalyticRequest CreateRequestFromText(string text)
+        {
+            // Based on email thread with Arvind Krishnaa Jagannathan <arjagann@microsoft.com>
```
Comment (jackgerrits) [inline Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs]: >// Based on email thread with Arvind Krishnaa Jagannathan <arjagann@microsoft.com> [](start = 12, length = 82)

Can you explain why this check in place?

**Comment by marco-rossi29** [comment]
I'd like to merge this asap to restore featurization for complex. What is the status?

**Comment by lokitoth** [comment]
Latest changes edit the history to remove the nuget and versioning changes which are premature. It also introduces constants to clarify where the magic numbers 10240 and 10240 / 2 are coming from.

File: Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs [hunk 5]
```diff
@@ -24,6 +24,26 @@ static CognitiveServiceTextAnalytics()
...
+            if (text.Length >= Constants.MaxRequestSizeUtf16)
+            text = text.Substring(0, Constants.MaxRequestSizeUtf16);
```
Comment (jackgerrits) [inline Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs]: >text = text.Substring(0, Constants.MaxRequestSizeUtf16); [](start = 12, length = 56)

braces or indentation

File: Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs [hunk 6]
```diff
@@ -42,27 +62,16 @@ public static Task<HttpResponseMessage> Run(HttpRequestMessage req, TraceWriter
                     var text = textBuilder.ToString();
 
                     // Based on email thread with Arvind Krishnaa Jagannathan <arjagann@microsoft.com>
-                    if (text.Length >= 10240 / 2)
-                        text = text.Substring(0, 10240 / 2);
+                    if (text.Length >= Constants.MaxRequestSizeUtf16)
+                        text = text.Substring(0, Constants.MaxRequestSizeUtf16);
```
Comment (jackgerrits) [inline Crawl/CognitiveServiceTextAnalytics/CognitiveServiceTextAnalytics.cs]: duplicated check above?

**Pull request merged**